language/bytecode-verifier/src/abstract_state.rs

```diff
@@ -346,96 +364,93 @@ impl AbstractState {
-        state1.locals.keys().any(|idx| {
-            state1.locals[idx].value.is_value()
-                && state1.is_local_borrowed(*idx)
-                && !state2.locals.contains_key(idx)
-        })
+    fn iter_locals(&self) -> impl Iterator<Item = LocalIndex> {
+        self.max_local.into_iter().flat_map(|max| 0..=max)
     }
 
-    fn split_locals(
-        locals: &BTreeMap<LocalIndex, TypedAbstractValue>,
-        values: &mut BTreeMap<LocalIndex, Kind>,
-        references: &mut BTreeMap<LocalIndex, Nonce>,
-    ) {
-        for (idx, abs_type) in locals {
-            match abs_type.value {
-                AbstractValue::Reference(nonce) => {
-                    references.insert(idx.clone(), nonce);
+    pub fn join_(&self, other: &Self) -> Option<Self> {
+        checked_precondition!(self.is_canonical() && other.is_canonical());
+        checked_precondition!(self.next_id == other.next_id);
+        checked_precondition!(self.max_local == other.max_local);
+        checked_precondition!(self.frame_root == other.frame_root);
+        // A join failure (returning None) occurs in each of the following situations:
+        // - a resource is available along one path but not the other
+        // - a value is borrowed along one path but unavailable along the other
+        let max_local = self.max_local;
+        let mut locals = BTreeMap::new();
+        let mut self_removed = BTreeSet::new();
+        let mut other_removed = BTreeSet::new();
+        for idx in self.iter_locals() {
+            let self_value = self.locals.get(&idx);
+            let other_value = other.locals.get(&idx);
+            match (self_value, other_value) {
+                // join error, a resource is available along one path but not the other
+                (None, Some(v)) | (Some(v), None) if v.value.is_possibly_resource() => return None,
+
+                // A reference is assigned on one side, but not the other. Release
+                (Some(v), None) => {
+                    if let AbstractValue::Reference(id) = &v.value {
+                        self_removed.insert(*id);
+                    }
                 }
-                AbstractValue::Value(kind) => {
-                    values.insert(idx.clone(), kind);
+                (None, Some(v)) => {
+                    if let AbstractValue::Reference(id) = &v.value {
+                        other_removed.insert(*id);
+                    }
                 }
+
+                // The local has a value on each side, add it to the state
+                (Some(v1), Some(v2)) => {
+                    checked_verify!(v1 == v2);
+                    checked_verify!(!locals.contains_key(&idx));
+                    locals.insert(idx, v1.clone());
+                }
+
+                (None, None) => (),
             }
         }
+
+        let borrow_graph = {
+            let mut self_graph = self.borrow_graph.clone();
```

Comment:
Since these are both cloned, I think you could actually get rid of the `self_removed` / `other_removed` sets and do the removal directly at 393 and 398?

Reply:
Sure I'll do that. I was just trying to keep things a bit separate